integration/internal/network/l2disco_linux.go

```diff
@@ -0,0 +1,210 @@
+package network
+
+// Collect and decode broadcast ARP and unsolicited ICMP6 Neighbour Advertisement messages...
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net"
+	"net/netip"
+	"slices"
+	"testing"
+	"time"
+
+	"github.com/docker/docker/internal/nlwrap"
+	"golang.org/x/sys/unix"
+	"gotest.tools/v3/assert"
+)
+
+// TimestampedPkt has a Data slice representing a packet, ReceivedAt is a timestamp
+// set after the packet was received in user-space.
+type TimestampedPkt struct {
+	ReceivedAt time.Time
+	Data       []byte
+}
+
+// CollectBcastARPs collects broadcast ARPs from interface ifname.
+// It returns a stop function, to stop collection and return a slice of collected packets (with
+// timestamps added when they were received in userspace).
+func CollectBcastARPs(t *testing.T, ifname string) (stop func() []TimestampedPkt) {
+	sd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_DGRAM|unix.SOCK_CLOEXEC, int(htons(unix.ETH_P_ARP)))
+	assert.NilError(t, err)
+	assert.Assert(t, sd >= 0)
+
+	link, err := nlwrap.LinkByName(ifname)
+	assert.NilError(t, err)
+
+	err = unix.Bind(sd, &unix.SockaddrLinklayer{
+		Protocol: htons(unix.ETH_P_ARP),
+		Pkttype:  unix.PACKET_BROADCAST,
+		Ifindex:  link.Attrs().Index,
+	})
+	assert.NilError(t, err)
+
+	return collectPackets(t, sd, "ARP")
+}
+
+// From https://github.com/mdlayher/packet/blob/f9999b41d9cfb0586e75467db1c81cfde4f965ba/packet_linux.go#L238-L248
+func htons(i uint16) uint16 {
+	var bigEndian [2]byte
+	binary.BigEndian.PutUint16(bigEndian[:], i)
+	return binary.NativeEndian.Uint16(bigEndian[:])
+}
+
+// CollectICMP6 collects ICMP6 packets sent to the all nodes address.
+// It returns a stop function, to stop collection and return a slice of collected packets (with
+// timestamps added when they were received in userspace).
+func CollectICMP6(t *testing.T, ifname string) (stop func() []TimestampedPkt) {
+	sd, err := unix.Socket(unix.AF_INET6, unix.SOCK_RAW|unix.SOCK_CLOEXEC, unix.IPPROTO_ICMPV6)
+	assert.NilError(t, err)
+	assert.Assert(t, sd >= 0)
+
+	link, err := nlwrap.LinkByName(ifname)
+	assert.NilError(t, err)
+
+	mreq := &unix.IPv6Mreq{
+		Interface: uint32(link.Attrs().Index),
+	}
+	copy(mreq.Multiaddr[:], net.IPv6linklocalallnodes)
+	err = unix.SetsockoptIPv6Mreq(sd, unix.IPPROTO_IPV6, unix.IPV6_JOIN_GROUP, mreq)
+	assert.NilError(t, err)
+
+	return collectPackets(t, sd, "ICMP6")
+}
+
+func collectPackets(t *testing.T, sd int, what string) (stop func() []TimestampedPkt) {
+	err := unix.SetsockoptTimeval(sd, unix.SOL_SOCKET, unix.SO_RCVTIMEO, &unix.Timeval{Sec: 1})
+	assert.NilError(t, err)
+
+	stopC := make(chan struct{})
+	stoppedC := make(chan struct{})
+	var pkts []TimestampedPkt
+	go func() {
+		defer close(stoppedC)
+		defer unix.Close(sd)
+		for {
+			buf := make([]byte, 50)
+			n, err := unix.Read(sd, buf)
+			select {
+			case <-stopC:
+				return
+			default:
+			}
```

Comment:
Shouldn't we check for the stop *before* actually doing the Read syscall? It's probably more expensive than the rest of the loop (checking error and appending to the slice).

```suggestion
			select {
			case <-stopC:
				return
			default:
			}
			n, err := unix.Read(sd, buf)
```

Reply:
Ahh, right... it's not clear, so I've added some comments to the function that might help.

It's a blocking `Read` with a timeout ... so, if no packet is received for one second it'll return an `EWOULDBBLOCK`.

When that happens, there's a good chance the caller has asked for packet collection to stop (by calling the function returned by `collectPackets`). In that case, this packet-collecting goroutine just needs to exit ... there's probably no packet to process and append to `pkts` - but it doesn't matter, that packet's not needed.

So, checking `stopC` before the read would be more-or-less equivalent - because the `EWOULDBLOCK` would just cause a `continue`, back to the top of the loop where, `stopC` would be checked ... but there's no need to go back around the loop.